### PR TITLE
add Pairing Tuple (->) and Pairing Product Coproduct instances

### PR DESCRIPTION
**Pull request by colin-lamed**
Not sure if the scope of the submodules is just to support the web-framework (Proact module), but the CofreeT and type-class Pairing are useful to other projects, not using the web-framework. Maybe worth extracting into their own libraries?

### CODE REVIEW

**Comment by alvart** [comment]
Yes, the scope of all the submodules was to support Proact. Specifically, because there are alternatives to these modules in more standard libraries that for one or another reason I decided not to use:

- **[Data.Functor.Pairing](https://pursuit.purescript.org/packages/purescript-pairing/4.0.0/docs/Data.Functor.Pairing)** (from your pull request): This seems the standard way to handle Functor Pairings. I decided not to use this package because I needed the Functor relationship to be defined as an instance instead of as an ordinary function. While this makes the code "prettier" by forcing the compiler to guess the type-class instance, there is a price to pay as there is one for all type-classes in Purescript/Haskell, maybe that's way Phil decided not to go down this route.
- **[Control.Monad.Free.Trans](https://pursuit.purescript.org/packages/purescript-freet/3.0.0)** An implementation of Free Monad Transformers that is optimized, I would assume. I decided not to use it just because I implemented the Cofree Comonad Transformer module and thought it would look cool to have both duals coming from the same package, heh.

The only module that had not "standard" implementations was the Cofree Comonad Transformer which is strange seeing as there is an implementation for [Cofree](https://pursuit.purescript.org/packages/purescript-free/4.2.0/docs/Control.Comonad.Cofree). Whatever the reason, I can see this module being useful for other libraries but I'd argue that the most proper and official place for it would be inside the purescript-freet package next to its dual. I didn't add it there because I wanted to finish my library ASAP and didn't want to bother myself into going through all the process that's required to modify and version a library with much larger impact.

And lastly, I made up new type classes for "Hoist" and "Interpret" transformers to make my Proact Components behave nicer in certain scenarios (unit testing) but I'm not sure these would be helpful in other projects (potentially) and if so, which "official package" would be their right place to be (maybe a new one?).

So basically, I'm not sure if it'd be worth it to extract the more abstract modules of Proact into their own packages. What do you think?

**Comment by colin-lamed** [comment]
I agree that purescript-freet would be the best place for CofreeT. I think it would be worth submitting your CofreeT contribution there.

I also think it could be interesting to have a type-class Pairing as an alternative to the purescript-pairing data type approach, maybe as it's own package. Although I think I may have seen some limitations to this approach in purescript - e.g. Overlappable instance warnings, and "possibly infinite instance" errors.

I was also interested in the FreeT CofreeT pairing, but without a "standard" CofreeT, it's not clear where that would go. Probably it should go in purescript-freet, with dependency on purescript-pairing.
